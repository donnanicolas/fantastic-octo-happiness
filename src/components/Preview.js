// @flow
import React, { PureComponent } from 'react';
import Sound from 'react-sound';

import Button from './Button';

type Props = {
  track: Track,
};

type State = {
  playing: boolean
};


class Preview extends PureComponent {

  state: State = {
    playing: false,
  };
  props: Props;

  handlePlayingChanged = (playing: boolean) => () =>
    this.setState(state => ({ ...state, playing }));

  render() {
    const { playing } = this.state;
    const { track } = this.props;
    const { previewUrl } = track;

    const playStatus = playing
      ? Sound.status.PLAYING
      : Sound.status.STOPPED;

    return (
      <div>
        <Sound
          url={previewUrl}
          playStatus={playStatus}
          onFinishedPlaying={this.handlePlayingChanged(false)}
        />
        <Button
          primary={!playing} 
          danger={playing}
          onClick={this.handlePlayingChanged(!playing)} 
        >
          {playing
            ? 'Stop'
            : 'Play'
          }
        </Button>
      </div>

    )


  }
}

export default Preview;
