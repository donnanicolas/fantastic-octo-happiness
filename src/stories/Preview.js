// @flow
import React from 'react';
import { storiesOf } from '@storybook/react';

import Preview from '../components/Preview';

import tracks from '../fixtures/tracks';

const [track] = tracks;
const noPreviewTrack = { ...track, previewUrl: undefined };

const addStories = () => {
  storiesOf('Preview', module)
    .add('with previewUrl', () => <Preview track={track}/>)
    .add('without previewUrl', () => <Preview track={noPreviewTrack} />);
};

export default addStories;