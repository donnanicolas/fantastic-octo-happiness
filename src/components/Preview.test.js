import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Preview from './Preview';

import tracks from '../fixtures/tracks';

const [track] = tracks;
const localFileTrack = { ...track, previewUrl: '../fixtures/moose.mp3' };
const noPreviewTrack = { ...track, previewUrl: undefined };

it('renders a preview button with previewUrl', () => {
  const tree = renderer.create(
    <Preview track={localFileTrack} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders a preview button with NO previewUrl', () => {
  const tree = renderer.create(
    <Preview track={noPreviewTrack} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders a preview button and clicks it once', () => {
  const wrapper = mount(
    <Preview track={localFileTrack} />
  )

  const button = wrapper.find('button');
  wrapper.find('button').simulate('click');

  expect(button.hasClass('btn-danger')).toBe(true);
  expect(button.text()).toBe('Stop');
});

it('renders a preview button and clicks it twice', () => {
  const wrapper = mount(
    <Preview track={localFileTrack} />
  )

  const button = wrapper.find('button');
  wrapper.find('button').simulate('click');
  wrapper.find('button').simulate('click');

  expect(button.hasClass('btn-primary')).toBe(true);
  expect(button.text()).toBe('Preview');
});

it('renders a preview button with out preview', () => {
  const wrapper = mount(
    <Preview track={noPreviewTrack} />
  )

  const button = wrapper.find('button');
  wrapper.find('button').simulate('click');

  expect(button.hasClass('btn-primary')).toBe(true);
  expect(button.text()).toBe('No Preview Available');
});

