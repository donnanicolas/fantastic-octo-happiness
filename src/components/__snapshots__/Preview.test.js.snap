// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a preview button with NO previewUrl 1`] = `
<button
  className="jEtlWH btn btn-primary btn-block"
  disabled={true}
  onClick={[Function]}
  onTouchEnd={[Function]}
  type="button"
>
  No Preview Available
  <span
    className="glyphicon glyphicon-play"
  />
</button>
`;

exports[`renders a preview button with previewUrl 1`] = `
<button
  className="jEtlWH btn btn-primary btn-block"
  disabled={false}
  onClick={[Function]}
  onTouchEnd={[Function]}
  type="button"
>
  Preview
  <span
    className="glyphicon glyphicon-play"
  />
</button>
`;
